File: tax_service.py
"""Tax calculation service."""
from typing import List
from models import TaxRequest, TaxResponse, ItemTaxDetail, Address


class TaxService:
    """TaxService handles tax calculation logic."""
    
    def calculate_tax(self, req: TaxRequest) -> TaxResponse:
        """Calculate tax for the given request.
        
        Args:
            req: The tax calculation request
            
        Returns:
            The tax calculation response
            
        Raises:
            ValueError: If the request is invalid
        """
        self._validate_request(req)
        
        # Get tax rate based on location
        tax_rate = self._get_tax_rate_for_location(req.address)
        jurisdiction = self._get_tax_jurisdiction(req.address)
        
        item_details: List[ItemTaxDetail] = []
        subtotal = 0.0
        total_tax = 0.0
        
        # Calculate tax for each item
        for item in req.items:
            item_subtotal = item.price * item.quantity
            item_tax = item_subtotal * tax_rate
            item_total = item_subtotal + item_tax
            
            detail = ItemTaxDetail(
                item_id=item.id,
                item_name=item.name,
                price=item.price,
                quantity=item.quantity,
                subtotal=self._round_to_two_decimals(item_subtotal),
                tax_rate=self._round_to_two_decimals(tax_rate * 100),  # Convert to percentage
                tax_amount=self._round_to_two_decimals(item_tax),
                total_amount=self._round_to_two_decimals(item_total)
            )
            
            item_details.append(detail)
            subtotal += item_subtotal
            total_tax += item_tax
        
        response = TaxResponse(
            address=req.address,
            items=item_details,
            subtotal=self._round_to_two_decimals(subtotal),
            total_tax=self._round_to_two_decimals(total_tax),
            grand_total=self._round_to_two_decimals(subtotal + total_tax),
            tax_jurisdiction=jurisdiction
        )
        
        return response
    
    def _validate_request(self, req: TaxRequest) -> None:
        """Validate the tax calculation request.
        
        Args:
            req: The request to validate
            
        Raises:
            ValueError: If the request is invalid
        """
        if not req.address.state:
            raise ValueError("state is required")
        if not req.address.zipcode and not req.address.postal_code:
            raise ValueError("zipcode is required")
        if not req.items:
            raise ValueError("at least one item is required")
        
        for i, item in enumerate(req.items):
            if item.price < 0:
                raise ValueError(f"item {i} has invalid price")
            if item.quantity <= 0:
                raise ValueError(f"item {i} has invalid quantity")
    
    def _get_tax_rate_for_location(self, address: Address) -> float:
        """Get tax rate based on the US state.
        
        Rates are approximate and based on combined state and average local rates.
        
        Args:
            address: The address to get the tax rate for
            
        Returns:
            The tax rate as a decimal (e.g., 0.0852 for 8.52%)
        """
        state = address.state.upper()
        
        # US state sales tax rates (approximate combined rates)
        tax_rates = {
            "AL": 0.0913,  # Alabama
            "AK": 0.0176,  # Alaska
            "AZ": 0.0831,  # Arizona
            "AR": 0.0947,  # Arkansas
            "CA": 0.0850,  # California
            "CO": 0.0763,  # Colorado
            "CT": 0.0635,  # Connecticut
            "DE": 0.0000,  # Delaware - No sales tax
            "FL": 0.0705,  # Florida
            "GA": 0.0733,  # Georgia
            "HI": 0.0444,  # Hawaii
            "ID": 0.0602,  # Idaho
            "IL": 0.0868,  # Illinois
            "IN": 0.0700,  # Indiana
            "IA": 0.0694,  # Iowa
            "KS": 0.0865,  # Kansas
            "KY": 0.0600,  # Kentucky
            "LA": 0.0952,  # Louisiana
            "ME": 0.0550,  # Maine
            "MD": 0.0600,  # Maryland
            "MA": 0.0625,  # Massachusetts
            "MI": 0.0600,  # Michigan
            "MN": 0.0744,  # Minnesota
            "MS": 0.0707,  # Mississippi
            "MO": 0.0824,  # Missouri
            "MT": 0.0000,  # Montana - No sales tax
            "NE": 0.0694,  # Nebraska
            "NV": 0.0823,  # Nevada
            "NH": 0.0000,  # New Hampshire - No sales tax
            "NJ": 0.0663,  # New Jersey
            "NM": 0.0779,  # New Mexico
            "NY": 0.0852,  # New York
            "NC": 0.0698,  # North Carolina
            "ND": 0.0696,  # North Dakota
            "OH": 0.0723,  # Ohio
            "OK": 0.0897,  # Oklahoma
            "OR": 0.0000,  # Oregon - No sales tax
            "PA": 0.0634,  # Pennsylvania
            "RI": 0.0700,  # Rhode Island
            "SC": 0.0744,  # South Carolina
            "SD": 0.0645,  # South Dakota
            "TN": 0.0955,  # Tennessee
            "TX": 0.0820,  # Texas
            "UT": 0.0719,  # Utah
            "VT": 0.0624,  # Vermont
            "VA": 0.0575,  # Virginia
            "WA": 0.0920,  # Washington
            "WV": 0.0650,  # West Virginia
            "WI": 0.0543,  # Wisconsin
            "WY": 0.0536,  # Wyoming
        }
        
        # Default rate if state not recognized
        return tax_rates.get(state, 0.0700)
    
    def _get_tax_jurisdiction(self, address: Address) -> str:
        """Get the tax jurisdiction string.
        
        Args:
            address: The address to get the jurisdiction for
            
        Returns:
            The jurisdiction string
        """
        return f"{address.state}, USA"
    
    @staticmethod
    def _round_to_two_decimals(value: float) -> float:
        """Round a float to 2 decimal places.
        
        Args:
            value: The value to round
            
        Returns:
            The rounded value
        """
        return round(value, 2)

File: services/tax_service.go
package services

import (
	"fmt"
	"math/rand"
	"strings"
	"time"

	"github.com/vijayraghavareddy/tax-calculation/models"
)

// TaxService handles tax calculation logic
type TaxService struct {
	rand *rand.Rand
}

// NewTaxService creates a new instance of TaxService
func NewTaxService() *TaxService {
	source := rand.NewSource(time.Now().UnixNano())
	return &TaxService{
		rand: rand.New(source),
	}
}

// CalculateTax calculates tax for the given request
func (s *TaxService) CalculateTax(req *models.TaxRequest) (*models.TaxResponse, error) {
	if err := s.validateRequest(req); err != nil {
		return nil, err
	}

	// Get tax rate based on location
	taxRate := s.getTaxRateForLocation(&req.Address)
	jurisdiction := s.getTaxJurisdiction(&req.Address)

	var itemDetails []models.ItemTaxDetail
	var subtotal, totalTax float64

	// Calculate tax for each item
	for _, item := range req.Items {
		itemSubtotal := item.Price * float64(item.Quantity)
		itemTax := itemSubtotal * taxRate
		itemTotal := itemSubtotal + itemTax

		detail := models.ItemTaxDetail{
			ItemID:      item.ID,
			ItemName:    item.Name,
			Price:       item.Price,
			Quantity:    item.Quantity,
			Subtotal:    roundToTwoDecimals(itemSubtotal),
			TaxRate:     roundToTwoDecimals(taxRate * 100), // Convert to percentage
			TaxAmount:   roundToTwoDecimals(itemTax),
			TotalAmount: roundToTwoDecimals(itemTotal),
		}

		itemDetails = append(itemDetails, detail)
		subtotal += itemSubtotal
		totalTax += itemTax
	}

	response := &models.TaxResponse{
		Address:         req.Address,
		Items:           itemDetails,
		Subtotal:        roundToTwoDecimals(subtotal),
		TotalTax:        roundToTwoDecimals(totalTax),
		GrandTotal:      roundToTwoDecimals(subtotal + totalTax),
		TaxJurisdiction: jurisdiction,
	}

	return response, nil
}

// validateRequest validates the tax calculation request
func (s *TaxService) validateRequest(req *models.TaxRequest) error {
	if req.Address.State == "" {
		return fmt.Errorf("state is required")
	}
	if req.Address.ZipCode == "" && req.Address.PostalCode == "" {
		return fmt.Errorf("zipcode is required")
	}
	if len(req.Items) == 0 {
		return fmt.Errorf("at least one item is required")
	}

	for i, item := range req.Items {
		if item.Price < 0 {
			return fmt.Errorf("item %d has invalid price", i)
		}
		if item.Quantity <= 0 {
			return fmt.Errorf("item %d has invalid quantity", i)
		}
	}

	return nil
}

// getTaxRateForLocation returns a tax rate based on location
// Routes to appropriate country-specific tax calculation
func (s *TaxService) getTaxRateForLocation(address *models.Address) float64 {
	country := strings.ToUpper(address.Country)
	
	// Route to country-specific tax calculation
	if country == "IN" || country == "INDIA" {
		return s.getIndianStateTaxRate(address)
	}
	
	// Default to US tax calculation
	return s.getUSTaxRate(address)
}

// getUSTaxRate returns a tax rate based on the US state
// Rates are approximate and based on combined state and average local rates
func (s *TaxService) getUSTaxRate(address *models.Address) float64 {
	state := strings.ToUpper(address.State)

	// US state sales tax rates (approximate combined rates)
	var baseRate float64
	switch state {
	case "AL": // Alabama
		baseRate = 0.0913
	case "AK": // Alaska
		baseRate = 0.0176
	case "AZ": // Arizona
		baseRate = 0.0831
	case "AR": // Arkansas
		baseRate = 0.0947
	case "CA": // California
		baseRate = 0.0850
	case "CO": // Colorado
		baseRate = 0.0763
	case "CT": // Connecticut
		baseRate = 0.0635
	case "DE": // Delaware
		baseRate = 0.0000 // No sales tax
	case "FL": // Florida
		baseRate = 0.0705
	case "GA": // Georgia
		baseRate = 0.0733
	case "HI": // Hawaii
		baseRate = 0.0444
	case "ID": // Idaho
		baseRate = 0.0602
	case "IL": // Illinois
		baseRate = 0.0868
	case "IN": // Indiana
		baseRate = 0.0700
	case "IA": // Iowa
		baseRate = 0.0694
	case "KS": // Kansas
		baseRate = 0.0865
	case "KY": // Kentucky
		baseRate = 0.0600
	case "LA": // Louisiana
		baseRate = 0.0952
	case "ME": // Maine
		baseRate = 0.0550
	case "MD": // Maryland
		baseRate = 0.0600
	case "MA": // Massachusetts
		baseRate = 0.0625
	case "MI": // Michigan
		baseRate = 0.0600
	case "MN": // Minnesota
		baseRate = 0.0744
	case "MS": // Mississippi
		baseRate = 0.0707
	case "MO": // Missouri
		baseRate = 0.0824
	case "MT": // Montana
		baseRate = 0.0000 // No sales tax
	case "NE": // Nebraska
		baseRate = 0.0694
	case "NV": // Nevada
		baseRate = 0.0823
	case "NH": // New Hampshire
		baseRate = 0.0000 // No sales tax
	case "NJ": // New Jersey
		baseRate = 0.0663
	case "NM": // New Mexico
		baseRate = 0.0779
	case "NY": // New York
		baseRate = 0.0852
	case "NC": // North Carolina
		baseRate = 0.0698
	case "ND": // North Dakota
		baseRate = 0.0696
	case "OH": // Ohio
		baseRate = 0.0723
	case "OK": // Oklahoma
		baseRate = 0.0897
	case "OR": // Oregon
		baseRate = 0.0000 // No sales tax
	case "PA": // Pennsylvania
		baseRate = 0.0634
	case "RI": // Rhode Island
		baseRate = 0.0700
	case "SC": // South Carolina
		baseRate = 0.0744
	case "SD": // South Dakota
		baseRate = 0.0645
	case "TN": // Tennessee
		baseRate = 0.0955
	case "TX": // Texas
		baseRate = 0.0820
	case "UT": // Utah
		baseRate = 0.0719
	case "VT": // Vermont
		baseRate = 0.0624
	case "VA": // Virginia
		baseRate = 0.0575
	case "WA": // Washington
		baseRate = 0.0920
	case "WV": // West Virginia
		baseRate = 0.0650
	case "WI": // Wisconsin
		baseRate = 0.0543
	case "WY": // Wyoming
		baseRate = 0.0536
	default:
		// Default rate if state not recognized
		baseRate = 0.0700
	}

	return baseRate
}

// getIndianStateTaxRate returns GST rate for Indian states
// India uses a centralized GST system with standard rates: 5%, 12%, 18%, 28%
// Using 18% as default (most common rate for goods and services)
func (s *TaxService) getIndianStateTaxRate(address *models.Address) float64 {
	state := strings.ToUpper(address.State)

	// Indian state GST rates (CGST + SGST = Total GST)
	// Most goods/services fall under 18% GST (9% CGST + 9% SGST)
	var gstRate float64
	switch state {
	case "MH", "MAHARASHTRA":
		gstRate = 0.18 // 18% GST
	case "KA", "KARNATAKA":
		gstRate = 0.18 // 18% GST
	case "TN", "TAMIL NADU", "TAMILNADU":
		gstRate = 0.18 // 18% GST
	case "DL", "DELHI":
		gstRate = 0.18 // 18% GST
	case "GJ", "GUJARAT":
		gstRate = 0.18 // 18% GST
	case "WB", "WEST BENGAL", "WESTBENGAL":
		gstRate = 0.18 // 18% GST
	case "RJ", "RAJASTHAN":
		gstRate = 0.18 // 18% GST
	case "UP", "UTTAR PRADESH", "UTTARPRADESH":
		gstRate = 0.18 // 18% GST
	case "MP", "MADHYA PRADESH", "MADHYAPRADESH":
		gstRate = 0.18 // 18% GST
	case "AP", "ANDHRA PRADESH", "ANDHRAPRADESH":
		gstRate = 0.18 // 18% GST
	case "TS", "TELANGANA":
		gstRate = 0.18 // 18% GST
	case "BR", "BIHAR":
		gstRate = 0.18 // 18% GST
	case "HR", "HARYANA":
		gstRate = 0.18 // 18% GST
	case "PB", "PUNJAB":
		gstRate = 0.18 // 18% GST
	case "KL", "KERALA":
		gstRate = 0.18 // 18% GST
	case "OR", "ODISHA", "ORISSA":
		gstRate = 0.18 // 18% GST
	case "JH", "JHARKHAND":
		gstRate = 0.18 // 18% GST
	case "AS", "ASSAM":
		gstRate = 0.18 // 18% GST
	case "CT", "CHHATTISGARH":
		gstRate = 0.18 // 18% GST
	case "UK", "UTTARAKHAND":
		gstRate = 0.18 // 18% GST
	case "HP", "HIMACHAL PRADESH", "HIMACHALPRADESH":
		gstRate = 0.18 // 18% GST
	case "JK", "JAMMU AND KASHMIR", "JAMMUANDKASHMIR":
		gstRate = 0.18 // 18% GST
	case "GA", "GOA":
		gstRate = 0.18 // 18% GST
	default:
		// Default GST rate for unrecognized states
		gstRate = 0.18
	}

	return gstRate
}

// getTaxJurisdiction returns the tax jurisdiction string
func (s *TaxService) getTaxJurisdiction(address *models.Address) string {
	country := strings.ToUpper(address.Country)
	
	if country == "IN" || country == "INDIA" {
		return fmt.Sprintf("%s, India (GST)", address.State)
	}
	
	return fmt.Sprintf("%s, USA", address.State)
}

// roundToTwoDecimals rounds a float64 to 2 decimal places
func roundToTwoDecimals(value float64) float64 {
	return float64(int(value*100+0.5)) / 100
}
